core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java

```diff
@@ -4107,6 +4113,93 @@ public static List arrayExcept(List list1, List list2) {
     return new ArrayList<>(result);
   }
 
+  /** Support the ARRAY_INSERT function. */
+  public static @Nullable List arrayInsert(List baselist, Object pos, Object val) {
+    if (baselist == null || pos == null) {
+      return null;
+    }
+    int posInt = (int) pos;
+    Object[] baseArray = baselist.toArray();
+    if (posInt == 0) {
+      throw new IllegalArgumentException("The index 0 is invalid. "
+          + "An index shall be either < 0 or > 0 (the first element has index 1)");
+    }
+
+    boolean usePositivePos = posInt > 0;
+
+    if (usePositivePos) {
+      int newArrayLength = Math.max(baseArray.length + 1, posInt);
+
+      if (newArrayLength > MAX_ARRAY_LENGTH) {
+        throw new IndexOutOfBoundsException(
+            String.format(Locale.ROOT, "The new array length %s exceeds the allowed limit.",
+                newArrayLength));
+      }
+
+      Object[] newArray = new Object[newArrayLength];
+
+      int posIndex = posInt - 1;
+      if (posIndex < baseArray.length) {
+        System.arraycopy(baseArray, 0, newArray, 0, posIndex);
+        newArray[posIndex] = val;
+        System.arraycopy(baseArray, posIndex, newArray, posIndex + 1, baseArray.length - posIndex);
+      } else {
+        System.arraycopy(baseArray, 0, newArray, 0, baseArray.length);
+        newArray[posIndex] = val;
+      }
+
+      return Arrays.asList(newArray);
+    } else {
+      int posIndex = posInt;
+
+      boolean newPosExtendsArrayLeft = baseArray.length + posIndex < 0;
+
+      if (newPosExtendsArrayLeft) {
+        // special case, if the new position is negative but larger than the current array size
+        // place the new item at start of array, place the current array contents at the end
+        // and fill the newly created array elements in middle with a null
+        int newArrayLength = -posIndex + 1;
```

Comment:
same question -posIndex maybe having a number overflow exception? Maybe we can use:
`int newArrayLength = Integer.MAX_VALUE - (Integer.MAX_VALUE + posIndex);`
Then the next copy code will be :
`System.arraycopy(baseArray, 0, newArray, newArrayLength - baseArray.length + 1, baseArray.length);`

Reply:
thanks for reminding me. however, we can not use
`int newArrayLength = Integer.MAX_VALUE - (Integer.MAX_VALUE + posIndex);`
if posIndex = Integer.MIN_VALUE, it will be 
`int newArrayLength = Integer.MAX_VALUE + 1;` and overflow.

In fact, pos cannot exceed the maximum array limit, we can check in advance just like pos cannot be 0. and -pos will not overflow in subsequent code logic.

I have updated the PR and add overflow test cases. If you are free, PTAL.

Reply:
why not just compute using long values?

Reply:
> why not just compute using long values?

thanks for suggestions. the `posIndex` should < Integer.MAX_VALUE - 15, we can simply use Integer to cover it and without overflow. Otherwise, i think it's a good way to check the parameters in advance.